### ISSUE TITLE
Transform `ItemNotFoundException` to constraint violation

### ISSUE THREAD

**Issue by ihmels** [api-platform/core#5521]
If a dereferenceable IRI is used in a request and no resource exists under this IRI, an
`ItemNotFoundException` is thrown, which results in a response with status code 500 with the
following content:

```json
{
  "@context": "/api/contexts/Error",
  "@type": "hydra:Error",
  "hydra:title": "An error occurred",
  "hydra:description": "Internal Server Error"
}
```

The content of the request was:

```json
{
  "product": "/api/products/not_existing"
}
```

This response is not very helpful for the user of the API. Therefore it would be better if such an
error could be converted into a validation error:

```json
{
  "@context": "/api/contexts/ConstraintViolationList",
  "@type": "ConstraintViolationList",
  "hydra:title": "An error occurred",
  "hydra:description": "product: Item not found for \"/api/products/not_existing\".",
  "violations": [
    {
      "propertyPath": "product",
      "message": "Item not found for \"/api/products/not_existing\".",
      "code": null
    }
  ]
}
```

For this, we could use the mechanism by which type errors are collected during denormalization
(`PartialDenormalizationException`).

**Comment by soyuka** [comment]
Good question, I guess we could do something directly within deserialization? Not sure how PartialDenormalizationException works, nor how we transform exceptions to violations.

**Comment by soyuka** [comment]
closing as question

**Comment by ihmels** [comment]
This issue was fixed in #5844.

**Comment by karolymarton** [comment]
It's not working in API Platform 4.1, `ItemNotFoundException(\sprintf('Item not found for "%s".', $iri))` is thrown, but later it's changed to 'Unable to generate an IRI for the item of type ...'.
Also, _exception_to_status_ has no effect on this. Same code is still working under API Platform 3.3.
